File: Trees/BinaryTrees/NextPointerPopulation.java
package Trees.BinaryTrees;

import java.util.LinkedList;
import java.util.Queue;

public class NextPointerPopulation {
    static class Node {
        int data;
        Node left, right, next;

        public Node(int data) {
            this.data = data;
        }

        @Override
        public String toString() {
            return "Node{" +
                    "data=" + data +
                    '}';
        }
    }

    private static Node getInstance() {
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.left.left = new Node(4);
        root.left.right = new Node(6);
        root.right.right = new Node(7);
        return root;
    }
    // 1 ,2 ,3 ,4 ,5

    private static void populateNextPointers() {
        Node root = getInstance();
        Queue<Node> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            int currentLevelSize = q.size();
            for (int i = 0; i < currentLevelSize; i++) {
                Node currentNode = q.poll();
                Node nextNode = null;

                if (i + 1 == currentLevelSize) nextNode = null;
                else nextNode = q.peek();

                currentNode.next = nextNode;

                if (currentNode.left != null) q.add(currentNode.left);
                if (currentNode.right != null) q.add(currentNode.right);
            }
        }
    }

    //https://leetcode.com/problems/populating-next-right-pointers-in-each-node-i
    private static void populateNextPointersSpaceOptimized() {
        Node root = getInstance();

        Node startNode = root;
        while (startNode != null) {
            Node currentNode = startNode;
            while (currentNode != null) {
                if (currentNode.left != null) currentNode.left.next = currentNode.right;
                if (currentNode.right != null && currentNode.next != null)
                    currentNode.right.next = currentNode.next.left;

                currentNode = currentNode.next;
            }

            startNode = startNode.left;
        }
    }

    //https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii
    private static void populateNextPointers2SpaceOptimized(Node root) {
        Node startNode = root;
        while (startNode != null) {
            Node iteratorNode = startNode;
            while (iteratorNode != null) {

                iteratorNode = iteratorNode.next;
            }
        }
    }

    /***
     *                      1
     *                 2        3
     *             4     -   -       5
     *
     *
     */

    private static void mutate(TreeNode root, TreeNode parent, boolean isLeft) {
        if (root == null) return;

        TreeNode left = root.left;
        TreeNode right = root.right;

        mutate(left, root, true);
        if (parent == null) {
            root.right = parent;
        } else if (isLeft) {
            root.right = parent.right;
        } else {
            if (parent.right != null) root.right = parent.right.left;
            else root.right = null;
        }
        mutate(right, root, false);
    }


    private static void populateNextPointersDFS(Node root, Node parent, boolean isLeft) {
        if (root == null) return;

        if (isLeft && parent != null) {
            root.next = parent.right;
        }
        populateNextPointersDFS(root.left, root, true);
        populateNextPointersDFS(root.right, root, false);
    }

    public static void main(String[] args) {
//        populateNextPointers();
    }
}

File: Trees/BinaryTrees/NextPointerPopulation.cpp
#include <bits/stdc++.h>
using namespace std;

typedef struct TreeNode{
    int val;
    TreeNode* left,*right,*next;

    TreeNode(){
    }

    TreeNode(int val){
        this->val = val;
        left = right = next NULL;
    }
}TreeNode;

class Util{
	public:
    static TreeNode* getTree(){
        TreeNode* root = new TreeNode(10);
        root->left = new TreeNode(20);
        root->right = new TreeNode(30);

        root->left->left = new TreeNode(40);
        root->left->right = new TreeNode(50);
        root->left->right->left = new TreeNode(60);
        root->left->right->right = new TreeNode(70);
        return root;
    }

    static void populateNextPointer(TreeNode* root){
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()){
            int size = q.size();
            for(int i=0;i<size;i++){
                TreeNode* currentNode = q.front();
                q.pop();

                if(currentNode->left != nullptr)    q.push(currentNode->left);
                if(currentNode->right != nullptr)    q.push(currentNode->right);
                if(i == size - 1)   continue;


                TreeNode* nextNode = q.front();
                currentNode->next = nextNode;
            }
        }
    }

    //https://leetcode.com/problems/populating-next-right-pointers-in-each-node/
    static void populateNextPointerSpaceOptimized(TreeNode* root){
        TreeNode* rootIterator = root;
        while (rootIterator != nullptr){
            TreeNode* childIterator = rootIterator;
            while(childIterator != nullptr){
                if(childIterator->left != nullptr)  childIterator->left->next = childIterator->right;
                if(childIterator->right != nullptr && childIterator->next != nullptr)   childIterator->right->next = childIterator->next->left;

                childIterator = childIterator->next;
            }
            rootIterator = rootIterator->left;
        }
    }

    static void populateNextPointerSpaceOptimized2(TreeNode* root){
        TreeNode* rootIterator = root;
        TreeNode* prev = nullptr;

        while (rootIterator != nullptr){
            TreeNode* childIterator = rootIterator;
            while(childIterator != nullptr){

            }
            rootIterator = rootIterator->left;
            prev = nullptr;
        }
    }
};

int main(){

    return 0;
}